### ISSUE TITLE
Prefs performance

### ISSUE THREAD

**Issue by alexKatona** [B3nedikt/restring#74]
Hello there,

I am trying to improve the app startup time, I already make some parallelism but now I see that initiaizing Restring takes some time. Below is a simple benchmark of methods. I am using EncryptedSharedPreferences from jetpack security library in both LocaleProvider and StringRepository.

SPEEDTESTING    Restring    duration millis: 4
SPEEDTESTING    setupLocaleProvider    duration millis: 286
SPEEDTESTING    setupStringRepository    duration millis: 1159
SPEEDTESTING    ViewPump    duration millis: 0

```
    private fun setupStringRepository(context: Context) {
                CachedStringRepository(
                        SharedPrefsStringRepository { sharedPreferencesName ->
                            AppLocalePreferencesHelper.create(context, STORED_LOCALES_ALIAS, sharedPreferencesName)
                        }
                )
    }

```

One of the improvements could be not creating quantity and array preferences

![2020-12-07 13_39_19-app – LocaleManager kt  app arch-android-business](https://user-images.githubusercontent.com/14858450/101352166-00848e80-3892-11eb-8a5d-64003243929f.png)

I was also thinking to copy paste the SharedPrefsStringRepository and just remove stuff I dont need (quantity and arrays) which is basically a quick win. Have you thought about refactoring some parts of the library to paralelly create those rpeferences for example for each locale?

EDIT: 
those classes are internal so copy paste actually would not help if I want to keep most of the functionality as it was in lib:
![2020-12-07 13_54_19-app – EncryptedSharedPrefsStringRepository kt  app arch-android-business](https://user-images.githubusercontent.com/14858450/101353402-cd42ff00-3893-11eb-976e-342eb88601df.png)

and this is final

![2020-12-07 13_56_03-app – EncryptedSharedPrefsStringRepository kt  app arch-android-business](https://user-images.githubusercontent.com/14858450/101353480-f5caf900-3893-11eb-8bf5-b22b88b2649f.png)

**Comment by B3nedikt** [comment]
I am not sure if not creating quantity and string array really helps here, as shared preferences instances for these are only created if the application uses quantity strings or string arrays. The influence on start up time should be minimal.

But still if you want to optimize it, you can do something like this:

```kotlin
class EncryptedSharedPrefStringRepository(
        context: Context
) : MutableStringRepository {

    val delegate by lazy {
        CachedStringRepository(
                SharedPrefsStringRepository { sharedPreferencesName ->
                    // Create encrypted shared preferences here...
                }
        )
    }

    override val quantityStrings: MutableMap<Locale, MutableMap<String, Map<PluralKeyword, CharSequence>>>
        get() = mutableMapOf()

    override val stringArrays: MutableMap<Locale, MutableMap<String, Array<CharSequence>>>
        get() = mutableMapOf()

    override val strings: MutableMap<Locale, MutableMap<String, CharSequence>>
        get() = delegate.strings

    override val supportedLocales: Set<Locale>
        get() = delegate.supportedLocales
}
```

Basically the SharedPrefsStringRepository is not intended to be used with inheritance, but using delegation.

Parallelly creating shared preferences is at the moment not planned, I think the better approach would be to create the preferences in a lazy fashion whenever a new locale is needed. This is to some point already implemented, a few performance improvements around this will be in the next version of restring. This together with some other small optimizations will improve start up time quite a bit. I think this is the more generic approach, most apps will will not need all strings in all locales available at application start. And with 30 somewhat locales it will be quite the load even if they are created in parallel. A kind of parallel mode for the edge cases where it actually makes sense too load everything in parallel may be an option though. 

That aside your start-up time is quite long even for the current version, way higher than in my own tests with many locales & strings, maybe this is related to the usage of EncryptedSharedPreferences?

I write when I have finished the new version in this issue, probably around next week.

**Comment by alexKatona** [comment]
Thanks for the explanation and info. Yeah the issue here is the EncryptedSharedPreferences for sure which I unfortunatelly have to use. I used the delegated repository and the results are quite good

I tested the whole initialization which previously took around 2100ms, now it takes 1000ms after skipping the quantity and array prefs

//first run after clearing data of app
SPEEDTESTING    init    duration millis: 1549

//prefs already created 
SPEEDTESTING    init    duration millis: 1020
SPEEDTESTING    init    duration millis: 1006

**Comment by B3nedikt** [comment]
Interesting, I did not expect that this would improve the performance this much. I just released version 5.1.1 of restring, it should have a much lower startup time especially if your app has a lot of strings and different languages. Please try it out, I think it may make it also faster for your use case :)
